lib/Backend/Lower.cpp

```diff
@@ -19584,26 +19612,66 @@ Lowerer::GenerateArgOutForStackArgs(IR::Instr* callInstr, IR::Instr* stackArgsIn
 
-    subInstr = IR::Instr::New(Js::OpCode::Sub_I4, ldLenDstOpnd, ldLenDstOpnd, IR::IntConstOpnd::New(1, TyMachReg, func),func);
-    callInstr->InsertBefore(subInstr);
-    this->m_lowererMD.EmitInt4Instr(subInstr);
+    // Note: This loop iteratively adds instructions in two locations; in the block
+    // of branches that jump to the "load elements to argOuts" instructions, and in
+    // the the block of load elements to argOuts instructions themselves.
 
-    nthArgument = IR::IndirOpnd::New(stackArgs, ldLenDstOpnd, TyMachReg, func);
-    ldElemDstOpnd = IR::RegOpnd::New(TyMachReg,func);
-    const IR::AutoReuseOpnd autoReuseldElemDstOpnd2(ldElemDstOpnd, func);
-    ldElem = IR::Instr::New(Js::OpCode::LdElemI_A, ldElemDstOpnd, nthArgument, func);
-    callInstr->InsertBefore(ldElem);
-    GenerateFastStackArgumentsLdElemI(ldElem);
+    // 4 to denote this is 4th register after this, callinfo & function object
+    // INT_ARG_REG_COUNT is the number of parameters passed in int regs
+    uint current_reg_pass =
+#if defined(_M_IX86)
+        // We get a compilation error on x86 due to assiging a negative to a uint
+        // TODO: don't even define this function on x86 - we Assert(false) anyway there.
+        0;
+#else
+        INT_ARG_REG_COUNT - 4;
+#endif
 
-    argout = IR::Instr::New(Js::OpCode::ArgOut_A_Dynamic, func);
-    argout->SetSrc1(ldElemDstOpnd);
-    callInstr->InsertBefore(argout);
-    this->m_lowererMD.LoadDynamicArgument(argout, 4); //4 to denote this is 4th register after this, callinfo & function object
+    do
+    {
+        // If we're on this pass we know we have to do at least one of these, so skip
+        // the branch if we're on the last one.
+        if (current_reg_pass != INT_ARG_REG_COUNT - 4)
+        {
+            IR::LabelInstr* loadBlockLabel = IR::LabelInstr::New(Js::OpCode::Label, func);
+            IR::Instr* branchToBlock = IR::BranchInstr::New(Js::OpCode::BrEq_I4, loadBlockLabel, ldLenDstOpnd, IR::IntConstOpnd::New(current_reg_pass + 1, TyInt8, func), func);
+            endLoop->InsertAfter(branchToBlock);
+            callInstr->InsertBefore(loadBlockLabel);
+        }
+
+        // TODO: We can further optimize this with a GenerateFastStackArgumentsLdElemI that can
+        // handle us passing along constant argument references and encode them into the offset
+        // instead of having to use an IndirOpnd; this would allow us to save a few bytes here,
+        // and reduce register pressure a hair
+
+        // stemp.var = LdImm current_reg_pass
+        IR::RegOpnd* localTemp = IR::RegOpnd::New(TyInt32, func);
+        // We need to make it a tagged int because GenerateFastStackArgumentsLdElemI asserts if
+        // it is not.
+        localTemp->SetValueType(ValueType::GetTaggedInt());
+        const IR::AutoReuseOpnd autoReuseldElemDstOpnd3(localTemp, func);
+        this->InsertMove(localTemp, IR::IntConstOpnd::New(current_reg_pass, TyInt8, func, true), callInstr);
```

Comment:
>InsertMove [](start = 14, length = 10)

why do you need this? couldn't you set an IntConstOpnd as the indexOpnd of nthArgument or current_reg_pass*TyPtr as its offset?

Reply:
I looked into going either of these paths initially, but there's unfortunately a problem with each. IndirOpnds currently require the indexOpnd to be a RegOpnd, preventing the use of IntConstOpnds there, and I wasn't sure that all the paths in GenerateFastStackArgumentsLdElemI would respect an existing offset (iirc some overwrite it without checking for a previous value). It's definitely something that can be improved upon, just that I didn't see a way to do so without expanding the scope of the commit (something I'm a bit hesitant to do due to the schedule).